fix(ssi): resolve images for every mutation

betterengineering opened the pull request:
### What does this PR do?
This commit fixes an issue where we only resolve image tags during startup and not on every pod mutation by ensuring the resolution happens just before use.

### Motivation
We've added gradual rollout support for Single Step Instrumentation so that language libraries are released in a gradual fashion in https://github.com/DataDog/datadog-agent/pull/39915. This was missed during code review and caught during testing.

### Describe how you validated your changes
I tested this using [injector-dev](https://github.com/DataDog/injector-dev):
```
injector-dev apply -f dev.yaml --profile staging --build
```

<details>
  <summary>dev.yaml</summary>

```yaml
helm:
  apps:
    - name: gradual-rollout-test
      namespace: application
      values:
        env:
          - name: DD_TRACE_DEBUG
            value: "true"
          - name: DD_APM_INSTRUMENTATION_DEBUG
            value: "true"
        image:
          repository: registry.ddbuild.io/ci/injector-dev/python
          tag: 2cd78ded
        podLabels:
          language: python
          tags.datadoghq.com/env: local
        service:
          port: "8080"
  versions:
    agent: 7.69.1
    cluster_agent:
      version: 7.69.1
      build: {}
    injector:
      version: 0.44.0
  config:
    clusterAgent:
      env:
        - name: DD_REMOTE_CONFIGURATION_ENABLED
          value: "true"
    datadog:
      site: "datad0g.com"
      apm:
        instrumentation:
          enabled: true
          targets:
            - name: python
              podSelector:
                matchLabels:
                  language: python
              ddTraceVersions:
                python: default
```

</details>

### Additional Notes
We will need this backported to `7.71.x`

betterengineering committed:
fix(ssi): resolve images for every mutation

This commit fixes an issue where we only resolve image tags during
startup and not on every pod mutation by ensuring the resolution happens
just before use.

betterengineering committed:
Add release note

TonyCTHsu commented on pkg/clusteragent/admission/mutate/autoinstrumentation/language_versions.go:
```diff
@@ -224,6 +230,14 @@ func (i libInfo) initContainers(v version) []initContainer {
+		log.Debugf("Resolving image %s/%s:%s", i.registry, i.repository, i.tag)
+		image, ok := i.imageResolver.Resolve(i.registry, i.repository, i.tag)
+		if ok {
+			i.image = image.FullImageRef
```
This should apply the same pattern 
https://github.com/DataDog/datadog-agent/pull/40762/files#diff-75a60dbf5546cab4c0507ad10f27369dae56c856c860443129282f7dba73282cL60-R60

betterengineering commented on pkg/clusteragent/admission/mutate/autoinstrumentation/language_versions.go:
```diff
@@ -166,9 +168,13 @@ func (l language) defaultLibVersion() string {
-	image   string
+	ctrName       string // empty means all containers
+	lang          language
+	image         string
```
The image key is the full image. ie `gcr.io/datadoghq/dd-lib-python-init:abc123`. The repository is just the repository: `gcr.io/datadoghq`.

betterengineering committed:
Merge branch 'main' into mark.spicer/fix-image-resolution

betterengineering committed:
Move release note

erikayasuda commented on dev.yaml:
```diff
@@ -0,0 +1,42 @@
+helm:
```
nit: Should we remove this?

betterengineering replied:
Woah, definitely

erikayasuda commented on releasenotes-dca/notes/fix-image-resolution-gradual-rollout-5199027c8888de72.yaml:
```diff
@@ -0,0 +1,5 @@
+---
```
nit: Adel made a similar comment in my other PR, but do we need the release note if this is before the release is cut?

betterengineering replied:
Removed

erikayasuda reviewed (approved):
Just some nits, thank you for the fix!

adel121 committed:
Makes image resolver a dependency of mutator instead of libinfo (#40806)

### What does this PR do?

Makes image resolver a dependency of mutator instead of libInfo

This PR is opened as a suggestion to a dev branch. It will be merged on
the dev branch, not on main.

betterengineering committed:
Address PR feedback

adel121 committed:
fix lint

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/injector.go:
```diff
@@ -150,9 +150,16 @@ func (i *injector) requirements() libRequirement {
-	key: "admission.datadoghq.com/apm-inject.version",
-	do:  infallibleFn(injectorWithImageTag),
+var injectorVersionAnnotationExtractorFunc = func(imageResolver ImageResolver) annotationExtractor[injectorOption] {
+	injectorVersionAnnotationExtractor := annotationExtractor[injectorOption]{
```
nit: indentation/can just return the annotationExtractor without an intermediate value

stanistan commented on pkg/clusteragent/admission/mutate/autoinstrumentation/language_versions.go:
```diff
@@ -51,18 +52,18 @@ func (l language) libInfo(ctrName, image string) libInfo {
-		image = l.libImageName(registry, version)
-	} else {
-		image = resolvedImage.FullImageRef
+func (l language) libInfoWithResolver(ctrName, registry string, version string) libInfo {
```
I think this was called out elsewhere, but `libInfoWithResolver` doesn't accept a resolver anymore :)

stanistan reviewed (approved):
A few notes, but LGTM overall, nice catch on the bug.
